src/unix/signal.c

```diff
@@ -234,20 +236,21 @@ static int uv__signal_register_handler(int signum, int oneshot) {
   if (oneshot)
     sa.sa_flags |= SA_RESETHAND;
 
-  /* XXX save old action so we can restore it later on? */
-  if (sigaction(signum, &sa, NULL))
+  if (sigaction(signum, &sa, &sa_old))
     return UV__ERR(errno);
 
+  uv__sigactions[signum] = sa_old;
+
   return 0;
 }
 
 
 static void uv__signal_unregister_handler(int signum) {
   /* When this function is called, the signal lock must be held. */
-  struct sigaction sa;
-
-  memset(&sa, 0, sizeof(sa));
-  sa.sa_handler = SIG_DFL;
+  struct sigaction sa = uv__sigactions[signum];
+  if (!sa.sa_handler) {
+    sa.sa_handler = SIG_DFL;
+  }
```

Comment:
Style:
```suggestion
  struct sigaction sa;

  sa = uv__sigactions[signum];
  if (sa.sa_handler == 0)
    sa.sa_handler = SIG_DFL;
```
Substance: I'm not 100% sure this check is correct everywhere.

On Linux and the BSDs, `SIG_DFL == 0`, so this check is basically a no-op and can be removed. But if another platform defines e.g. `SIG_IGN == 0`, then the check is wrong because it changes the handler from SIG_IGN to SIG_DFL.

Probably easiest to resolve for now by adding a `STATIC_ASSERT(0 == SIG_DFL)` unless you have a better idea.

Reply:
@bnoordhuis I thought about it in the start of the work, but decided not to consider this corner case. So, if you noticed this corner case too, I took it into account. I'v pushed updates.